core/src/com/treode/async/misc/DirectDeque.scala

```diff
@@ -0,0 +1,95 @@
+package com.treode.async.misc
+
+import java.lang.System
+import scala.reflect.ClassTag
+
+/**
+ * DirectDeque is a queue (FIFO ordering) that supports
+ * constant time access to any element in the queue
+ */
+private class DirectDeque [M] (init_capacity: Int = 16) (implicit
+   mtag: ClassTag[M]
+) {
+  private var front = 0
+  private var back = front
+
+  private var array = new Array[M](init_capacity)
+
+  def size = {
+    if (front == back) {
+      0
+    } else if (front < back) {
+      back-front
+    } else {
+      (array.length-front)+back
+    }
+  }
+
+  def isEmpty() : Boolean = {
+    front == back
+  }
+
+ /**
+  * Adds element to the end of this DirectDeque
+  */
+  def enqueue (element: M) {
+    array(back) = element
```

Comment:
The code that's there now has (back + 1) % array.length __twice__. You can make the val next to clear that up.